src/apispec/ext/marshmallow/openapi.py

```diff
@@ -110,35 +110,29 @@ def resolve_nested_schema(self, schema):
         return self.get_ref_dict(schema_instance)
 
     def schema2parameters(
-        self,
-        schema,
-        *,
-        default_in="body",
-        name="body",
-        required=False,
-        description=None
+        self, schema, *, location, name="body", required=False, description=None
     ):
         """Return an array of OpenAPI parameters given a given marshmallow
-        :class:`Schema <marshmallow.Schema>`. If `default_in` is "body", then return an array
+        :class:`Schema <marshmallow.Schema>`. If `location` is "body", then return an array
         of a single parameter; else return an array of a parameter for each included field in
         the :class:`Schema <marshmallow.Schema>`.
 
+        In OpenAPI 3, only "query", "header", "path" or "cookie" are allowed for the location
+        of parameters. "requestBody" is used when fields are in the body.
+
         https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#parameterObject
         """
-        openapi_default_in = __location_map__.get(default_in, default_in)
-        if self.openapi_version.major < 3 and openapi_default_in == "body":
-            prop = self.resolve_nested_schema(schema)
-
+        openapi_location = __location_map__.get(location, location)
```

Comment:
Ideally I would like to remove this mapping and rely on the calling code to know the appropriate OpenAPI locations.

Reply:
Yes, we should investigate this. I don't think I need this location, as my calling code could enter the correct value.

Indeed, this could, and probably should, be done in user code.

There are people relying on it, though. See #517.

Reply:
I opened #587.